### ISSUE TITLE
Table Character Set Disregarded upon SQL dump / Export

### ISSUE THREAD

**Issue by ye** [sequelpro/sequelpro#2321]
Another issue was discovered regarding charset/encoding: currently version of `sequelpro` seems to disregard the table's charset/encoding and using `utf8` charset regardless the table setting. In other words, if I have a table that's set to use `utf8mb4` charset the export SQL file will still **blindly** setting client connection to `utf8` charset.

For instance: `test_default` table was set with 
` DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci` 
but the exported SQL dump file still sets client connection with 
`/*!40101 SET NAMES utf8 */;`  

The consequence of this bug is that if I have data content outside of `utf8` charset, then this export is broken as it cannot represent those characters and will be replaced by `?`.

```sql
# ************************************************************
# Sequel Pro SQL dump
# Version 4499
#
# http://www.sequelpro.com/
# https://github.com/sequelpro/sequelpro
#
# Host: 127.0.0.1 (MySQL 5.5.38-0ubuntu0.14.04.1)
# Database: emoji_test
# Generation Time: 2015-11-10 21:14:08 +0000
# ************************************************************


/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;


# Dump of table test_default
# ------------------------------------------------------------

CREATE TABLE `test_default` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `foo` varchar(250) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `sticky_priority` enum('HIGH','MEDIUM','LOW') COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;




/*!40111 SET SQL_NOTES=@OLD_SQL_NOTES */;
/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
/*!40014 SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS */;
/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;

```

**Comment by dmoagx** [comment]
There are actually two issues here:
* On export: Sequel Pro requests utf8 and MySQL will return the non-bmp chars as `?`
* On import: Sequel Pro declares utf8 and MySQL will mangle the data before inserting.

What makes this issue difficult though, is, that I don't know for which other charsets this might happen and I can't find any resource that explains compatibility of the other ~38 charsets in mysql to utf8, either.

Mixing different charsets in one export file would introduce new risks and difficulties, too.
Returning everything as hex-encoded data would be safe, but completely illegible.

One option might be to request the data using the `binary` charset and then locally try to convert it using libicu limited to roundtrip conversion, and if that fails dump the string as hex-encoded instead.

**Comment by ye** [comment]
@dmoagx thanks for the analysis - though I am not quite understand.

Isn't this a simple issue of matching import/export charset with the corresponding table's charset designation?

I understand in the edge case where I can have a table A charset with `utf8` and table B with `gb2312` and exporting both will post a dilemma of choosing a single charset to export (hence the conversions). But I don't think Sequel Pro should decide and choose one for the user. It should prompt for user to decide - just like a git merge conflict - present the options let user decide, after all it's the users who get them into this mixing charset situation.

In any case, any fix (convert or let user decide) would be better off than the current broken situation IMMO. No offense.

**Comment by dmoagx** [comment]
The main problem here is, how do we detect when charset issues happen.
For example:
```
mysql> select * from tbl1;
+----+-------+
| id | value |
+----+-------+
|  5 | 🍯      |
+----+-------+
1 rows in set (0,00 sec)

mysql> show warnings;
Empty set (0,00 sec)

mysql> set names latin1;
Query OK, 0 rows affected (0,00 sec)

mysql> select * from tbl1;
+----+-------+
| id | value |
+----+-------+
|  5 | ?     |
+----+-------+
1 row in set (0,01 sec)

mysql> show warnings;
Empty set (0,00 sec)
```
Since I used the wrong charset in the second query MySQL just returned "?" instead of the character, but I don't have a way to check for that.

Likewise on import: 
We could export data with `utf8mb4`, but that was only added in MySQL 5.5.3.
If you were to import such a file on a older MySQL version it would just throw out the additional characters without the user ever noticing. This can cause rather evil creeping data corruption.

AFAIR the only thing that prevents that is the STRICT SQL mode, but that also includes other stuff we don't want.
```
mysql> set names utf8;
Query OK, 0 rows affected (0,00 sec)

mysql> set sql_mode='STRICT_ALL_TABLES';
Query OK, 0 rows affected (0,01 sec)

mysql> insert into tbl1(value) values('👾');
ERROR 1366 (HY000): Incorrect string value: '\xF0\x9F\x91\xBE' for column 'value' at row 1
```

Or we'd have to check the warnings after every query. Then at least Sequel Pro could show charset issues on import:
```
mysql> set names utf8;
Query OK, 0 rows affected (0,00 sec)

mysql> insert into tbl1(value) values('👾');
Query OK, 1 row affected, 1 warning (0,02 sec)

mysql> show warnings;
+---------+------+------------------------------------------------------------------------+
| Level   | Code | Message                                                                |
+---------+------+------------------------------------------------------------------------+
| Warning | 1366 | Incorrect string value: '\xF0\x9F\x91\xBE' for column 'value' at row 1 |
+---------+------+------------------------------------------------------------------------+
1 row in set (0,00 sec)
```

**Comment by dmoagx** [comment]
I've added code in 7d4ec977bfdc0fea0af962d3859ba6d39c3c3a2d which makes utf8mb4 the preferred export charset.
That should handle most of the common use cases, but won't solve the problem for characters that are not roundtrip-compatible with Unicode at all (if those exist, which I can't seem to figure out).

**Comment by ye** [comment]
I don't think uniformly set charset to `utf8mb4` would solve the import/export issue. 

IMMO, a better approach would be to respect whatever the origin charset was and set the import/export charset accordingly. Perhaps you have some other difficulties / issues that I am not aware off. Oh well.

**Comment by rowanbeentje** [comment]
@ye Agreed, but that's a more complex fix: the file would then have to use a mix of charsets within itself and be read purely as binary, so that wouldn't then be very compatible with other programs.  If the whole file is set to utf8mb4 so it's always readable in an editor, but the charset is changed for each set of tables within it, then there's be transcoding problems... could switch to a directory of files each with its own encoding but that will likely make some people unhappy too!

**Comment by dmoagx** [comment]
Mixing multiple charsets in one file is a sure-fire way to get problems.
Another part of the problem is that MySQL supports encodings on a per-column level.
The best bet is to use utf8mb4 for the file as that has the widest range of supported characters.

My idea would be to use read everything from the DB as binary, locally try to convert it to utf8mb4 and if that's not possible in a roundtrip compatible way output it in Hex and cast it to the original encoding.

The problem here is with encodings such as `geostd8` which don't even have any proper standard defining them (we could copy the unicode mapping from mysql though)

**Comment by ye** [comment]
Yes, I agree with the disagree: it certainly wasn't a trivial fix :) 

Since you mentioned, binary (hex encoded) seems to be a safe lowest common denominator if all else failed.

> My idea would be to read everything from the DB as binary, locally try to convert it to utf8mb4 and if that's not possible in a roundtrip compatible way (we can enforce that with libicu) output it in Hex and cast it to the original encoding.

I understand the difficulties of detecting codecs especially in the situation of "mixed charsets". If what you meant above is to encode mix charsets strings to binary and then re-decoding the bytes into unicdoe string then I think it may be doable. 

It would be good to test it against a couple test cases.

**Comment by wouterds** [comment]
Any update on this? Still not working in current stable version.

**Comment by dmoagx** [comment]
What *exactly* is not working?

**Comment by wouterds** [comment]
@dmoagx: The exact thing described in the original post. When exporting a table with `utf8mb4` charset it emoji becomes "?" in the sql dump. I just tried the nightly version and here it does seem to work.

**Comment by dmoagx** [comment]
Ah yes, we didn't include the partial fix in 1.1.x because some users complained about incompatibilities.

**Comment by abhibeckert** [comment]
So should we flag this as `fixed awaiting release`?

**Comment by dmoagx** [comment]
It's not exactly fixed, that would be what I described in my comment https://github.com/sequelpro/sequelpro/issues/2321#issuecomment-161024522 above.
But always using `utf8mb4` when it's available seems to be the best workaround.

**Comment by abhibeckert** [comment]
If we're already implementing the best workaround shouldn't we close the issue once it's released?

**Comment by filipsuk** [comment]
Any updates on this please? I have this problem in version 1.1.2.

**Comment by jorenvh1** [comment]
Anyone has found a fix for this?

**Comment by wouterds** [comment]
Bug was not present in nightly version afaicr. @jorenvh1
